pkg/zsio/zeek/flatten.go

```diff
@@ -0,0 +1,75 @@
+package zeek
+
+import (
+	"fmt"
+
+	"github.com/mccanne/zq/pkg/zeek"
+	"github.com/mccanne/zq/pkg/zson"
+	"github.com/mccanne/zq/pkg/zson/resolver"
+	"github.com/mccanne/zq/pkg/zval"
+)
+
+type Flattener struct {
+	mapper *resolver.Mapper
+}
+
+func NewFlatener() *Flattener {
+	return &Flattener{
+		mapper: resolver.NewMapper(resolver.NewTable()),
+	}
+}
+
+func (f *Flattener) Flatten(r *zson.Record) (*zson.Record, error) {
+	id := r.Descriptor.ID
+	d := f.mapper.Map(id)
+	if d == nil {
+		cols := flattenColumns(r.Type.Columns)
+		outRecord := zeek.LookupTypeRecord(cols)
+		d = f.mapper.Enter(id, outRecord)
```

Comment:
Why do we need to keep the flattened descriptor?

Reply:
Because it's better than recomputing for every record.  Computers are fast but it would go very measurably slower to not have some sort of approach like this.

Reply:
I just meant that in most cases the writer needs a flattened descriptor one time (to print a header), not for every record.

Reply:
The writer uses the descriptor's columns to do the recode on every record written.